File: AlgoLabs/3rd-Lab/J.java
import java.io.*;
import java.util.*;

public class Main {

    public static int getBit(int p, int i) {
        return (p >> i) & 1;
    }

    public static boolean match(int p, int p1, int n) {
        for (int i = 0; i < n - 1; i++) {
            if ((getBit(p, i) + getBit(p1, i) + getBit(p, i + 1) + getBit(p1, i + 1)) % 4 == 0) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        if (n > m) {
            int tmp = m;
            m = n;
            n = tmp;
        }
        int[][] dp = new int[m][1 << n];
        for (int i = 0; i < 1 << n; i++) {
            dp[0][i] = 1;
        }
        for (int i = 1; i < m; i++) {
            for (int p = 0; p < 1 << n; p++) {
                for (int p1 = 0; p1 < 1 << n; p1++) {
                    if (match(p, p1, n)) {
                        dp[i][p] += dp[i - 1][p1];
                    }
                }
            }
        }
        long res = 0L;
        for (int i = 0; i < 1 << n; i++) {
            res += dp[m - 1][i];
        }
        System.out.println(res);
    }
}

File: AlgoLabs/9th-Lab/J.kt
import java.util.*
import kotlin.collections.ArrayList

private var n: Int = 0
private var m: Int = 0
private val sc: Scanner = Scanner(System.`in`)
private lateinit var visited: Array<Boolean>
private lateinit var G: Array<HashSet<Int>>
private val ans: ArrayList<Int> = arrayListOf()
private lateinit var grandi: Array<Int>

private fun main() {
    n = sc.nextInt()
    m = sc.nextInt()
    G = Array(n) { hashSetOf<Int>() }
    visited = Array(n) { false }

    for (i in 0 until m) {
        val u = sc.nextInt() - 1
        val v = sc.nextInt() - 1
        G[u].add(v)
    }
    topsort()
    grandi = Array(n) { -1 }
    for (u in ans.reversed()) {

            val list: TreeSet<Int> = TreeSet()
            for (v in G[u]) list.add(grandi[v])
            grandi[u] = mex(list)
    }
    for (g in grandi){
        println(g)
    }
}

private fun topsort() {
    for (u in 0 until n) {
        if (!visited[u]) {
            dfs(u)
        }
    }
    ans.reverse()
}

private fun dfs(u: Int) {
    visited[u] = true
    for (v in G[u]) {
        if (!visited[v]) {
            dfs(v)
        }
    }
    ans.add(u)
}

private fun mex(list: TreeSet<Int>): Int {
    var min = 0
    for (c in list) {
        if (min == c) min++
        else break
    }
    return min;
}